cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -655,6 +633,8 @@ class PyListConverter : public ListConverter<T, PyConverter, PyConverterTrait> {
     return ValidateBuilder(this->list_type_);
   }
 
+  bool rewind_on_capacity_error() const override { return true; }
```

Comment:
Is this correct? The binary builder checks capacity before append, right?

Reply:
Yes, [ReserveData](https://github.com/apache/arrow/blob/1082c1c0501e534e333aa04cfb78f3091677a655/cpp/src/arrow/python/python_to_arrow.cc#L559) calls out to [ValidateOverflow](https://github.com/apache/arrow/blob/master/cpp/src/arrow/array/builder_binary.h#L310) before the `UnsafeAppend` call.

Reply:
Right - so we don't need to rewind, right?

Reply:
Right, but this is belongs to the `PyListConverter` (expand the diff).

Reply:
D'oh…sorry for the confusion.

Reply:
This fix is all about confusion :D